### PR TITLE
Consider tighter protobuf performance by removing array copies #1921

### PR DESCRIPTION
**Pull request by jleach4**
Small nit but see a lot of object creation in production.

### CODE REVIEW

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/InstrumentedTransaction.java [hunk 1]
```diff
@@ -130,6 +132,8 @@ public void clearRangeStartsWith(byte[] prefix) {
     public void mutate(MutationType opType, byte[] key, byte[] param) {
         underlying.mutate(opType, checkKey(key), param);
         /* Do we want to track each mutation type separately as well? */
+        //Thread.dumpStack();
+        System.out.println(String.format("Mutation {opType=%s, key=%s}", opType.name(), Base64.getEncoder().encodeToString(key)));
```
Comment (MMcM) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/InstrumentedTransaction.java:136]: Leftover debugging code?

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/indexes/AtomicMutationIndexMaintainer.java [hunk 2]
```diff
@@ -169,6 +169,7 @@ protected <M extends Message> CompletableFuture<Void> updateIndexKeys(@Nonnull f
                     state.transaction.mutate(MutationType.BYTE_MAX, key, param);
                 }
             } else {
+                System.out.println("mutation -> " + state.index.getType() + " : " + state.index);
```
Comment (MMcM) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/indexes/AtomicMutationIndexMaintainer.java:172]: More debugging?

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java [hunk 3]
```diff
@@ -459,6 +460,7 @@ private <M extends Message> void addRecordCount(@Nonnull RecordMetaData metaData
         final Transaction tr = ensureContextActive();
         Key.Evaluated subkey = metaData.getRecordCountKey().evaluateSingleton(rec);
         final byte[] keyBytes = getSubspace().pack(Tuple.from(RECORD_COUNT_KEY).addAll(subkey.toTupleAppropriateList()));
+        System.out.println("WTF?");
```
Comment (MMcM) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java:463]: This too.

**Review by MMcM — CHANGES_REQUESTED** [review]

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/indexes/AtomicMutationIndexMaintainer.java (hunk 2 above)
Comment (jleach4) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/indexes/AtomicMutationIndexMaintainer.java:172]: oops

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java (hunk 3 above)
Comment (jleach4) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java:463]: triple oops... wrong branch :(

**Review by MMcM — DISMISSED** [review]

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/cursors/FlatMapPipelinedCursor.java [hunk 4]
```diff
@@ -326,17 +326,17 @@ public byte[] toBytes() {
 
             if (innerContinuation.isEnd()) {
                 // This was the last of the inner cursor. Take continuation from outer after it.
-                builder.setOuterContinuation(ByteString.copyFrom(outerResult.getContinuation().toBytes()));
+                builder.setOuterContinuation(ZeroCopyByteString.wrap(outerResult.getContinuation().toBytes()));
```
Comment (alecgrieser) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/cursors/FlatMapPipelinedCursor.java:329]: I don't see a harm here in using a `ZeroCopyByteString`, but we should also consider using the `RecordCursorContinuation::toByteString` method on continuations in situations like these. The `toByteString` method is newer, and so not all of the continuations are updated, but for continuations that implement it, it should skip a copy (potentially more than one).

**Review by alecgrieser — APPROVED** [review]

**Pull request merged**